compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/KernelDispatch.cpp

```diff
@@ -460,55 +460,71 @@ static LogicalResult setRootConfigForPackPeel4LevelTilingPipeline(
     return linalgOp.emitOpError("failed to fetch m/n/k dims.");
   }
 
+  // Set attribute for number of input loops.
+  MLIRContext *context = entryPointFn.getContext();
+  unsigned numLoops = linalgOp.getNumLoops();
+  auto loopAttr = NumInputLoopsAttr::get(context, numLoops);
+  setNumInputLoopsAttr(linalgOp, loopAttr);
+
   AMDAIEDeviceModel deviceModel = getDeviceModel(targetDevice);
 
   // ------------------------------------------------------
   // --------------- Set packing config -------------------
   // ------------------------------------------------------
-  MLIRContext *context = entryPointFn.getContext();
-  unsigned numLoops = linalgOp.getNumLoops();
-
-  // Pack level => 1.
-  SmallVector<int64_t> packedSizesL0(numLoops, 0);
-  packedSizesL0[mDims.back()] = packPeelTiling.m0Pack;
-  packedSizesL0[nDims.back()] = packPeelTiling.n0Pack;
-  packedSizesL0[kDims.back()] = packPeelTiling.k0Pack;
-
-  // For matmul, transpose B matrix from [K N n k] to [N K k n]
-  // For matmul_transpose_b, we don't have to transpose the B matrix,
-  // since it is already [N K n k]
-  SmallVector<int64_t> transposePackIndices = {0, 1, 2};
-  // There is no corresponding unpack for the specified pack operation
-  // 0 is used when unpack is empty
-  SmallVector<bool> unpackEmpty = {false, false, true};
+  bool isBatchMatmul = isa<linalg::BatchMatmulOp>(linalgOp);
   SmallVector<int64_t> innerPermA = setInnerPermA(isMatmulTransposeA(linalgOp));
   SmallVector<int64_t> innerPermB = setInnerPermB(isMatmulTransposeB(linalgOp));
-  SmallVector<SmallVector<int64_t>> innerPerm = {
-      innerPermA, innerPermB, {0, 1}};
-  bool isBatchMatmul = isa<linalg::BatchMatmulOp>(linalgOp);
   SmallVector<int64_t> outerPermA =
       setOuterPermA(isMatmulTransposeA(linalgOp), isBatchMatmul);
   SmallVector<int64_t> outerPermB =
       setOuterPermB(isMatmulTransposeB(linalgOp), isBatchMatmul);
-  SmallVector<SmallVector<int64_t>> outerPerm = {outerPermA, outerPermB};
-  // Add outer permutation for unpack. NOTE: This currently fails for some
-  // tests in the AIR pipeline.
-  if (isa<linalg::BatchMatmulOp>(linalgOp)) {
-    outerPerm.push_back({0, 2, 1});
-  } else {
-    outerPerm.push_back({1, 0});
-  }
 
-  auto packingConfigLevel0Attr = getPackingConfigPackingLevelAttr(
-      context, packedSizesL0, transposePackIndices, unpackEmpty, innerPerm,
-      outerPerm);
+  SmallVector<int64_t> transposePackIndices;
+  SmallVector<bool> unpackEmpty;
+  SmallVector<SmallVector<int64_t>> innerPerm;
+  SmallVector<SmallVector<int64_t>> outerPerm;
+  SmallVector<PackingConfigPackingLevelAttr> packingConfigLevelsVal;
+
+  // Pack level => 1.
+  // The first level of packing is not needed if the number of loops is larger
```

Comment:
What is 'the number of loops' ?  How is this related to sizes of the SmallVector<unsigned, 2> vectors M, N, K, batch in your previous PR?

Reply:
It is the number of loops from the input linalg op. It equals to the total number of dims of batch,M,N,K.